sandstone/client/sandstone/components/filetreedirective/tests/filetree.directive.test.js

```diff
@@ -1,208 +1,421 @@
+        beforeEach(function() {
+              scope.sd = {
+                  noSelections: true,
+                  multipleSelections: false,
+                  dirSelection: false
+              };
+              var el = angular.element('<div sandstone-filetree tree-data="ctrl.treeData" leaf-level="file" selection-desc="sd"></div>');
+              element = $compile(el)(scope);
+              scope.$digest();
 
-    it('should return true if given filepath is expanded', function(){
-      isolateScope.treeData.expandedNodes = [dirs[0]];
-      expect(isolateScope.isExpanded(dirs[0].filepath)).toBeTruthy();
-    });
+              // Get isolate scope
+              isolateScope = element.isolateScope();
+          });
+        it('initializeFiletree', function(){
+            // Create spies
+            spyOn(isolateScope, 'updateFiletree');
+            spyOn(isolateScope, 'deletedFile');
+            spyOn(isolateScope, 'pastedFiles');
 
-    it('should return false if given filepath is not expanded', function(){
-      expect(isolateScope.isExpanded(dirs[0].filepath)).not.toBeTruthy();
-    });
+            expect(isolateScope.leafLevel).toBe("file");
+            expect(isolateScope.selectionDesc.noSelections).toBeTruthy();
+            expect(isolateScope.selectionDesc.multipleSelections).not.toBeTruthy();
+            expect(isolateScope.selectionDesc.dirSelected).not.toBeTruthy();
+            expect(isolateScope.treeData.filetreeContents.length).toBe(2);
+            // Refresh
+            rootScope.$emit('refreshFiletree');
+            expect(isolateScope.updateFiletree).toHaveBeenCalled();
+            // Deleted File
+            rootScope.$emit('deletedFile', dirs[0]);
+            expect(isolateScope.deletedFile).toHaveBeenCalled();
+            // Pasted File
+            rootScope.$emit('pastedFiles', dirs[0].filepath);
+            expect(isolateScope.pastedFiles).toHaveBeenCalled();
+        });
 
-    it('should describe the selection', function(){
-      var node = isolateScope.treeData.filetreeContents[0];
-      isolateScope.describeSelection(node, true);
-      expect(isolateScope.treeData.selectedNodes.length).toBe(1);
-      isolateScope.describeSelection(node, false);
-      expect(isolateScope.treeData.selectedNodes.length).toBe(0);
-    });
+        it('getDirContents', function(){
+            isolateScope.describeSelection(isolateScope.treeData.filetreeContents[0], true);
+            isolateScope.getDirContents(isolateScope.treeData.filetreeContents[0], true);
+            // First directory should have 3 files
+            expect(isolateScope.treeData.filetreeContents[0].children.length).toBe(3);
+        });
 
-    it('should update the filetree on getting a filtree:created_file event', function() {
-        spyOn(isolateScope, 'updateFiletree');
-        rootScope.$emit('filetree:created_file', {
-            filepath: '/tmp/some-file.txt'
+        it('getNodeFromPath: node exists', function() {
```

Comment:
Where possible, group test cases together so that each unit only has one associated test, and then pack the different cases inside.